orttraining/orttraining/python/training/onnxblock/onnxblock.py

```diff
@@ -70,7 +73,20 @@ def __call__(self, *args, **kwargs):
 
         output = self.build(*args, **kwargs)
 
-        self._model = onnx.shape_inference.infer_shapes(accessor._GLOBAL_ACCESSOR.model)
+        if accessor._GLOBAL_ACCESSOR.has_path:
+            onnx.save(self.base, TEMP_ONNX_PATH, save_as_external_data=True, all_tensors_to_one_file=True)
+
+            onnx.shape_inference.infer_shapes_path(TEMP_ONNX_PATH)
+            # shape inferenced model is saved to original path
+            self._model = onnx.load(TEMP_ONNX_PATH)
+
+            # clean-up temp files
+            if os.path.exists(TEMP_ONNX_PATH):
+                os.remove(TEMP_ONNX_PATH)
+            if os.path.exists(TEMP_ONNX_PATH):
+                os.remove(TEMP_ONNX_PATH)
```

Comment:
```suggestion
        with tempfile.TemporaryDirectory() as temp_dir:
            temp_model_path = temp_dir + "/model.onnx"
            onnx.save(self.base, temp_model_path, save_as_external_data=True, all_tensors_to_one_file=True)

            onnx.shape_inference.infer_shapes_path(temp_model_path)
            # shape inferenced model is saved to original path
            self._model = onnx.load(temp_model_path)
```

Probably not necessary to have a global for the temp filename, and tempfile.TemporaryDirectory can handle the cleanup for you.

Reply:
for visibility: the tempfile with TemporaryDirectory approach was causing a minor bug where ONNX was unable to find the external data file. 

Also, we realized that the ModelProto that is loaded and passed around does not contain the information from external data, only the metadata for where to find the external data, so the external data files generated by intermediary steps should be saved in some form.

Scott suggested an approach where we save to the original model path that's passed in, but in the case that the original model was saved with "all_tensors_to_one_file" set to True, subsequent save_model calls created a separate single data file named something different.

I'm going with the approach where the temp data files are retained until the very end of the generate_artifacts call then deleted, but open to other approaches / suggestions!